machineconfiguration/v1alpha1/types_machineconfigstate.go

```diff
@@ -0,0 +1,125 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigState describes the health of the Machines on the system
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineConfigStateSpec `json:"spec"`
+	// +optional
+	Status MachineConfigStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigState `json:"items"`
+}
+
+// MachineConfigStateSpec describes the type of State we are managing
+type MachineConfigStateSpec struct {
+	// configuration describes the object reference information of a MachineConfigState
+	Config MachineConfigStateConfig `json:"configuration"`
+}
+
+// MachineConfigStateStatus holds the reported information on a particular MachineConfigState
+type MachineConfigStateStatus struct {
+	corev1.ObjectReference `json:",inline"`
+	// conditions represents the latest available observations of current state.
+	// +listType=atomic
+	// +optional
+	Conditions []ProgressionCondition `json:"conditions"`
```

Comment:
You should use `metav1.Condition` and this should be a map rather than an atomic set. Please follow the conventions for conditions here so that standard tooling works (eg oc can wait for a particular condition)

It should look something like
```
    // conditions represent the observations of a MachineConfigState's current state.
    // Known .status.conditions.type are: "Available", "Progressing", and "Degraded"
    // +patchMergeKey=type
    // +patchStrategy=merge
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```

Reply:
Do I need to have `Available` `Progressing` and `Degraded` as the only values? that Really limits what I can do here.

Reply:
You can have whatever you want, based on previous discussion I reckon there will be quite a few more. They should just be strings somewhere in the API as consts so that we have a track/enumeration of them